### PR TITLE
fix(sql): allow UNION between IPv4 and VARCHAR/STRING columns

### LINKED ISSUE #5245: Assertion Error with STRING UNION IPv4
### To reproduce

Run this query:
```sql
select '1'::varchar as col from long_sequence(1)
union all
select '1'::long from long_sequence(1)
union all
select '127.0.0.1'::ipv4 from long_sequence(1);
```

It throws AssertionError:
```
2024-12-12T09:20:56.665726Z C i.q.c.h.p.JsonQueryProcessorState [150323855726] internal error [ex=
java.lang.AssertionError
	at io.questdb.griffin.SqlCodeGenerator.generateCastFunctions(SqlCodeGenerator.java:1920)
	at io.questdb.griffin.SqlCodeGenerator.generateSetFactory(SqlCodeGenerator.java:4947)
	at io.questdb.griffin.SqlCodeGenerator.generateUnionAllFactory(SqlCodeGenerator.java:5742)
	at io.questdb.griffin.SqlCodeGenerator.generateSetFactory(SqlCodeGenerator.java:4950)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:3167)
	at io.questdb.griffin.SqlCodeGenerator.generate(SqlCodeGenerator.java:496)
	at io.questdb.griffin.SqlCompilerImpl.generateSelectOneShot(SqlCompilerImpl.java:3279)
	at io.questdb.griffin.SqlCompilerImpl.generateSelectWithRetries(SqlCompilerImpl.java:418)
	at io.questdb.griffin.SqlCompilerImpl.compileUsingModel(SqlCompilerImpl.java:2053)
	at io.questdb.griffin.SqlCompilerImpl.compileInner(SqlCompilerImpl.java:1825)
	at io.questdb.griffin.SqlCompilerImpl.compile(SqlCompilerImpl.java:307)
	at io.questdb.cairo.pool.SqlCompilerPool$C.compile(SqlCompilerPool.java:136)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.compileAndExecuteQuery(JsonQueryProcessor.java:531)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.execute0(JsonQueryProcessor.java:222)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.onRequestComplete(JsonQueryProcessor.java:294)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientRecv(HttpConnectionContext.java:954)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientOperation(HttpConnectionContext.java:295)
	at io.questdb.cutlass.http.HttpServer.handleClientOperation(HttpServer.java:321)
	at io.questdb.cutlass.http.HttpServer$1.lambda$$0(HttpServer.java:115)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:199)
	at io.questdb.cutlass.http.HttpServer$1.run(HttpServer.java:119)
	at io.questdb.mp.Worker.run(Worker.java:152)
, q=`select '1'::varchar from long_sequence(1)
```

### QuestDB version:

8.2.2-SNAPSHOT, 6f1e6f36f864b6

### OS, in case of Docker specify Docker and the Host OS:

irrelevant

### File System, in case of Docker specify Host File System:

irrelevant

### Full Name:

Jaromir Hamala

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jerrinot**
Previously, attempting to UNION IPv4 columns with VARCHAR/STRING columns would throw an AssertionError. The SqlCodeGenerator now properly generates casting functions to VARCHAR/STRING for such operations.

Example:
```sql
select '1'::varchar as col from long_sequence(1)
union all
select '127.0.0.1'::ipv4 from long_sequence(1);
```

fixes #5245

### CODE REVIEW

**Review by amunra — APPROVED** [review]
👍

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 27 / 27 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|26|26|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 27 / 27 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|26|26|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -6137,39 +6208,6 @@ private Function toLimitFunction(
         return func;
     }
 
-    // UNION_CAST_MATRIX captures all the combinations of "left" and "right" column types
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:6140]: @jerrinot union cast matrix is generate by thus code, we should not be manually updating it and the code must remain in place commented out

Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:6140]: @bluestreak01 sees tests in `SqlCodeGeneratorTest`. This is a change from a @amunra's commit: Not only is the generation kept in place (and printed), but Adam also added a test to check that the stored matrix is the same as the one auto-generated at test-time.

Comment (amunra) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:6140]: TL;DR: The generation code and the matrix had _already_ diverged, I brought them back together and added tests to they don't diverge in the future.

Steps I took:
* I uncommented the code and marked it as `@TestOnly`.
* Then added a test to assert it's the same as the table that's there.
* It diverged significantly.
* I kept whatever was checked in the table and updated the generation code.
* The generation code now outputs a new table during tests should we add new types.

I did this in the `nd_arr` branch first (as part of test case failures fixes), then as I was asking about code in this area Jaromir spotted a bug and fixed it.

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 27 / 27 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|26|26|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**